### ISSUE TITLE
Add benchmarking within InSpec Tests

### ISSUE THREAD

**Issue by Schwad** [inspec/inspec#5003]
## 🎛 Description

**Choose one**: 🙋 feature request

Something we will look at and continue to look at is performance. Not only is it an important element for the development team, but is a critical component of improving scaling support. There are elements of this gem that may get called thousands or millions of times on larger runs, so micro-improvements of 1 or 2 milliseconds runtime could be a game changer.

RSpec and other libraries actually maintain full benchmark records in their core repository, some for various methods and others for running their core library elements at varying scales.

A similar approach would be beneficial to seeing our growth and weak points moving forward for InSpec.

This is not an automation issue, but a foundation for us to try this new framework. If implemented and used, we can revisit automation benefits in future (6-12+ months).

## 👍 Acceptance Criteria
* initialize new `benchmarks` directory
* include a descriptive readme explaining benchmarks
* start with a view basic benchmarks, as runnable ruby files that include text data on prior benchmarking runs
* add new benchmarks when putting together new performance improvement issues, so that we can measure our performance improvement gains

## 🔗 Related Links

* RSpec Benchmarks https://github.com/rspec/rspec-core/tree/master/benchmarks

Aha! Link: https://chef.aha.io/features/SH-2357

**Comment by james-stocks** [comment]
This is an awesome idea. I put the `Continuous Delivery` label on this to try to express that this is to give us benchmarking for our own dev/test/release and that it is not adding a benchmarking feature to InSpec

**Comment by Schwad** [comment]
Thanks! Still trying to get a handle on all the labels

**Comment by james-stocks** [comment]
One possible plan is to have expeditor run a job that does a couple of small invocations of InSpec e.g. `inspec --version` and `inspec exec test/fixture/profile/...` then send details of the timings, OS we ran on, and InSpec git ref to Chef's internal data store.

We could also look at a telemetry point that for `inspec exec`, but that is not this ticket.

**Comment by Schwad** [comment]
So although a benchmarking folder is not off the table for internals (say individual components of InSpec for development, etc.), it looks like our first step here will be as @james-stocks indicated to do this capture during CI run and send to Chef internal datastore. 

Having a think this morning on what the logical approach might be, and drew something up. Open to comments if others have a different preferable approach. 

I drew this up (poorly) on my machine but I'm thinking first, we simply run `time` against `inspec exec benchmark_profile.rb`. That profile being as simple as possible. Perhaps we run this five times and take the averages to avoid outlier runs.

Then, we `post` the data out of Expeditor. This hits an endpoint in AWS API Gateway, which is configured against a Lambda function. This Lambda function's job is simply to take POST requests and then put them into Redshift.

This is a basic example but yes we should probably do an exact git ref/version/OS as well. 

I mentioned about benchmarking the past but that is out of scope for this issue and something that can be looked at after this infrastructure is put in place. Another out-of-scope idea was to also run a test against a large (5000+ lines) profile to see if there are different experiences with performance depending on profile size.
![untitled (4)](https://user-images.githubusercontent.com/7865030/86338606-6ff2b400-bc4a-11ea-8287-0ffe6fb90082.png)

**Comment by clintoncwolfe** [comment]
FWIW, the telemetry PR #4874 does exactly all this already. It logs invocation duration to the company telemetry datastore, which is already setup and maintained for you. Logs local platform, inspec version, and several other interesting things.

**Comment by Schwad** [comment]
Hey @clintoncwolfe , thanks for this point! 

So in theory with that functionality we could add a rake task into the build process (say `bundle exec rake benchmark:basic_profile`) and would the DSL look something like:

```ruby
configure_telemeter(config)

telemetry_time_invocation("basic_profile") do
  shellout('whatever-way-we-end-up-agreeing-we-would-like-to-run-this-command-etc')
end
```

Or maybe I've constructed that incorrectly as we'll be composing the config after running our command 🤔 

This is definitely something to keep in mind 🙏

**Comment by Schwad** [comment]
I just had a good chat with @james-stocks about this issue and I think the approach we're going to look at is getting some sort of implementation of this in ASAP within expeditor manually, but keep an eye on telemetry for future changes.

**Comment by james-stocks** [comment]
Yes please, the ask here is to collect some rough data _now_ without being gated by the completion of telemetry. As discussed I believe we just need RelEng's advice on how to give Expeditor or BuildKite credentials to post to Redshift.
